odoo/addons/base/models/res_partner.py

```diff
@@ -274,8 +277,23 @@ def _compute_get_ids(self):
 
     @api.depends('is_company', 'parent_id.commercial_partner_id')
     def _compute_commercial_partner(self):
+        self.env.cr.execute("""
+        WITH RECURSIVE cpid(id, commercial_partner_id) AS (
+            SELECT id, id FROM res_partner
+            WHERE is_company OR parent_id IS NULL
+        UNION
+            SELECT p.id, cpid.commercial_partner_id
+            FROM res_partner p, cpid
+            WHERE NOT is_company and p.parent_id = cpid.id
+        )
+        SELECT * FROM cpid WHERE id =ANY(%s)
```

Comment:
I see a little problem with this approach. On a *sample* database with 1.5M partner records, this query takes 7 full seconds to execute even if `self.ids` is a singleton. [-> Explain plan](https://explain.depesz.com/s/TVLF)

Reply:
Is not a `ANY(VALUES(...))` matter?
https://www.datadoghq.com/blog/100x-faster-postgres-performance-by-changing-1-line

Reply:
@moylop260 nope, this bug has been resolved in PostgreSQL 9.3 (see update at the end of the article).

We can accelerate the query by only computing value on the tree of `self.ids`
```sql
WITH RECURSIVE tree(id,parent_id,is_company) AS (
    SELECT id, parent_id, is_company
      FROM res_partner
     WHERE id = ANY(%s)
     UNION
    SELECT p.id, p.parent_id, p.is_company
      FROM res_partner p, tree t
     WHERE p.id = t.parent_id
),
cpid(id, commercial_partner_id) AS (
    SELECT id, id
      FROM tree
     WHERE is_company
        OR parent_id IS NULL
     UNION
    SELECT p.id, cpid.commercial_partner_id
     FROM tree p, cpid
    WHERE NOT is_company
      AND p.parent_id = cpid.id
)
SELECT * FROM cpid WHERE id = ANY(%s);
```

Reply:
@odony can you check your test case with @KangOl's query?

Reply:
@xmo-odoo the bottom-up query from @KangOl works much better indeed, but it still needs to be slightly modified to consider cases where `is_company` is NULL (as was the case with the previous implementation, BTW).
This NULL case should be included in the test cases too, if they did not catch it.

Here's a proposition of modified version with a single CTE. It  covers the NULL `is_company` case and seems to be marginally faster (12ms instead of 67ms on 500 IDS in my 1.5M test database).
For a single ID both versions run in less than 0.5 ms.

```sql
WITH RECURSIVE cpid(id, parent_id, commercial_partner_id, final) AS (
    SELECT id, parent_id, id,       
           ((is_company IS NOT NULL AND is_company) OR
            parent_id IS NULL) as final
      FROM res_partner                                         
     WHERE id = ANY(%s)
  UNION                                   
    SELECT cpid.id, p.parent_id, p.id,
           ((is_company IS NOT NULL AND is_company) OR
            p.parent_id IS NULL) as final
      FROM res_partner p JOIN cpid ON (cpid.parent_id = p.id)
     WHERE NOT cpid.final
)                
SELECT cpid.id, cpid.commercial_partner_id
  FROM cpid
 WHERE final AND id = ANY(%s);

```

Reply:
@odony how about `coalesce(is_company, false)` instead of `is_company is not null and is_company`?

Reply:
Looking at it more, and after writing some tests, the special handling for is_company being null seems completely unnecessary in that version: since we're checking for is_company (and then final) being `true`, them being `null` or `false` makes no difference. Was probably an issue with @KangOl's version as it checks specifically for `not is_company`.

Reply:
> @odony how about `coalesce(is_company, false)` instead of `is_company is not null and is_company`?

Sure, equivalent and even shorter :)

```sql
WITH RECURSIVE cpid(id, parent_id, commercial_partner_id, final) AS (
    SELECT id, parent_id, id,       
           (COALESCE(is_company, false) OR parent_id IS NULL) as final
      FROM res_partner                                         
     WHERE id = ANY(%s)
  UNION                                   
    SELECT cpid.id, p.parent_id, p.id,
           (COALESCE(is_company, false) OR p.parent_id IS NULL) as final
      FROM res_partner p JOIN cpid ON (cpid.parent_id = p.id)
     WHERE NOT cpid.final
)                
SELECT cpid.id, cpid.commercial_partner_id
  FROM cpid
 WHERE final AND id = ANY(%s);

Reply:
> Looking at it more, and after writing some tests, the special handling for is_company being null seems completely unnecessary in that version: since we're checking for is_company (and then final) being `true`, them being `null` or `false` makes no difference. Was probably an issue with @KangOl's version as it checks specifically for `not is_company`.

I don't think we can spare the COALESCE because we'd still get _unknown_ branches in the comparison when `parent_id` is set, and thus skip result rows.

Let's say `company_id` is NULL and `parent_id` = 3:
```sql
SELECT (NULL OR 3 IS NULL) as final;
 final  
--------
 <NULL>
(1 row)
```
Do your tests include a row where `is_company` is NULL and `parent_id` is set? ;-)

Reply:
> Let's say company_id is NULL and parent_id = 3:

This yields final=null instead of final=false, but once through the final where the result's the same, and `null or true` returns true.

> Do your tests include a row where is_company is NULL and parent_id is set? ;-)

Yup.

Reply:
> > Let's say company_id is NULL and parent_id = 3:
> 
> This yields final=null instead of final=false, but once through the `where` the result's the same, and `null or true` returns true.

Well, the `WHERE NOT cpid.final` inside the recursive branch will fail to pick up a NULL, and the final result row will be missing entirely. You must doing something different ;-)

Here is an example where one result line is missing when there is no COALESCE:

```sql
# select id, parent_id, is_company from res_partner where id in (4823, 62413, 25525);
  id   | parent_id | is_company 
-------+-----------+------------
  4823 |     62413 | <NULL>
 25525 |    <NULL> | f
 62413 |    <NULL> | t
(3 rows)

openerp=# WITH RECURSIVE cpid(id, parent_id, commercial_partner_id, final) AS (
    SELECT id, parent_id, id,       
           (is_company OR parent_id IS NULL) as final
      FROM res_partner                                         
     WHERE id = ANY(ARRAY[4823,25525])
  UNION                                   
    SELECT cpid.id, p.parent_id, p.id,
           (is_company OR p.parent_id IS NULL) as final
      FROM res_partner p JOIN cpid ON (cpid.parent_id = p.id)
     WHERE NOT cpid.final
)                
SELECT cpid.id, cpid.commercial_partner_id
  FROM cpid
 WHERE final AND id = ANY(ARRAY[4823,25525]);
  id   | commercial_partner_id 
-------+-----------------------
 25525 |                 25525
(1 row)
```

The above is fixed by adding the necessary COALESCE, either when computing `final` or when testing `final` in the recursive WHERE:

```sql
openerp=# WITH RECURSIVE cpid(id, parent_id, commercial_partner_id, final) AS (
    SELECT id, parent_id, id,       
           (is_company OR parent_id IS NULL) as final
      FROM res_partner                                         
     WHERE id = ANY(ARRAY[4823,25525])
  UNION                                   
    SELECT cpid.id, p.parent_id, p.id,
           (is_company OR p.parent_id IS NULL) as final
      FROM res_partner p JOIN cpid ON (cpid.parent_id = p.id)
     WHERE NOT (COALESCE(cpid.final, false))
)                
SELECT cpid.id, cpid.commercial_partner_id
  FROM cpid
 WHERE final AND id = ANY(ARRAY[4823,25525]);
  id   | commercial_partner_id 
-------+-----------------------
 25525 |                 25525
  4823 |                 62413
(2 rows)
```